source/common/common/assert.cc

```diff
@@ -28,15 +32,76 @@ class ActionRegistrationImpl : public ActionRegistration {
   static std::function<void()> debug_assertion_failure_record_action_;
 };
 
+class EnvoyBugRegistrationImpl : public ActionRegistration {
+public:
+  EnvoyBugRegistrationImpl(std::function<void()> action) {
+    ASSERT(envoy_bug_failure_record_action_ == nullptr,
+           "An ENVOY_BUG action was already set. Currently only a single action is supported.");
+    envoy_bug_failure_record_action_ = action;
+    counters_.clear();
+  }
+
+  ~EnvoyBugRegistrationImpl() override {
+    ASSERT(envoy_bug_failure_record_action_ != nullptr);
+    envoy_bug_failure_record_action_ = nullptr;
+  }
+
+  static bool shouldLogAndInvoke(const char* filename, int line) {
+    const auto name = absl::StrCat(filename, ",", line);
+
+    // Increment counter, inserting first if counter does not exist.
+    absl::ReleasableMutexLock lock(&mutex_);
+    auto counter_value = ++counters_[name];
+    lock.Release();
```

Comment:
minor: If this was my code, I'd use MutexLock instead of ReleasableMutexLock:

```
int counter_value = 0;
{
    absl::MutexLock lock(&mutex_);
    counter_value = ++counters_[name];
}
```

It's harder to accidentally change the contents of the critical section with the RAII lock.